baseplate/queue_consumer.py

```diff
@@ -0,0 +1,201 @@
+    to the queue.
+
+    Next, the process registers a consumer that receives messages from
+    the queue and feeds them to the ``handler``.
+
+    The ``handler`` function must take 3 arguments:
+
+    * ``context``: a baseplate context
+    * ``message_body``: the text body of the message
+    * ``message``: :py:class:`kombu.message.Message`
+
+    The consumer will automatically ack each message after the handler method
+    exits. If there is an error in processing and the message must be retried
+    the handler should raise an exception to crash the process. This will
+    prevent the ack and the message will be requeued at the head of the queue.
+
+    :param baseplate.core.Baseplate baseplate: A baseplate instance for the
+        service.
+    :param kombu.Exchange exchange:
+    :param kombu.connection.Connection connection:
+    :param str queue_name: The name of the queue.
+    :param list routing_keys: List of routing keys.
+    :param handler: The handler method.
+
+    """
+    queues = []
+    for routing_key in routing_keys:
+        queue = Queue(
+            name=queue_name,
+            exchange=exchange,
+            routing_key=routing_key,
+        )
+        queues.append(queue)
+
+    logger.info("registering %s as a handler for %r", handler.__name__, queues)
+    kombu_consumer = KombuConsumer.new(connection, queues)
+
+    logger.info("waiting for messages")
+    while True:
+        context = ConsumerContext()
+        with baseplate.make_server_span(context, queue_name) as span:
+            message = kombu_consumer.get_message(span)
+            handler(context, message.body, message)
+            message.ack()
+
+
+class _ConsumerWorker(ConsumerMixin):
+    def __init__(self, connection, queues, work_queue):
+        self.connection = connection
+        self.queues = queues
+        self.work_queue = work_queue
+
+    def get_consumers(self, Consumer, channel):
+        return [Consumer(
+            queues=self.queues,
+            on_message=self.on_message,
+        )]
+
+    def on_message(self, message):
+        self.work_queue.put(message)
+
+    def get_message(self, block, timeout):
+        try:
+            return self.work_queue.get(block=block, timeout=timeout)
+        except queue.Empty:
+            return None
+
+class ConsumerContext(object):
+    pass
+
+
+class BaseKombuConsumer(object):
+    """Base object for consuming messages from a queue.
+
+    A worker process accepts messages from the queue and puts them in a local
+    work queue. The "real" consumer can then get messages with
+    :py:meth:`get_message` or :py:meth:`get_batch`. It is that consumer's
+    responsibility to ack or reject messages.
+
+    Can be used directly, outside of standard baseplate context.
+
+    """
+    def __init__(self, worker, worker_thread):
+        self.worker = worker
+        self.worker_thread = worker_thread
+
+    @classmethod
+    def new(cls, connection, queues):
```

Comment:
I'm not sure if this is what you had in mind for the factory, but it does do a better job of setting things up and isolating the worker to make things testable.

Reply:
Yes, this is great! 

I think this method needs a docstring so it'll get picked up by sphinx. It's currently missing from the docs.

Reply:
👍 I'm a fan of factory classmethods like this.